overlord/hookstate/ctlcmd/is_connected_test.go

```diff
@@ -0,0 +1,154 @@
+)
+
+type isConnectedSuite struct {
+	st          *state.State
+	mockHandler *hooktest.MockHandler
+}
+
+var _ = Suite(&isConnectedSuite{})
+
+func (s *isConnectedSuite) SetUpTest(c *C) {
+	s.st = state.New(nil)
+	s.mockHandler = hooktest.NewMockHandler()
+}
+
+var isConnectedTests = []struct {
+	args, stdout, stderr, error string
+	exitCode                    int
+}{{
+	args:  "is-connected",
+	error: "the required argument `<plug|slot>` was not provided",
+}, {
+	args: "is-connected plug1",
+}, {
+	args: "is-connected slot1",
+}, {
+	// reported as not connected because of undesired flag
+	args:     "is-connected plug2",
+	exitCode: 1,
+}, {
+	// reported as not connected because of hotplug-gone flag
+	args:     "is-connected plug3",
+	exitCode: 1,
+}, {
+	args:     "is-connected slot2",
+	exitCode: 1,
+}, {
+	args:     "is-connected foo",
+	exitCode: 1,
+}}
+
+func (s *isConnectedSuite) testIsConnected(c *C, context *hookstate.Context) {
+	s.st.Set("conns", map[string]interface{}{
+		"snap1:plug1 snap2:slot2": map[string]interface{}{},
+		"snap1:plug2 snap3:slot3": map[string]interface{}{"undesired": true},
+		"snap1:plug3 snap4:slot4": map[string]interface{}{"hotplug-gone": true},
+		"snap3:plug4 snap1:slot1": map[string]interface{}{},
+	})
+
+	s.st.Unlock()
+	defer s.st.Lock()
+
+	for _, test := range isConnectedTests {
+		stdout, stderr, err := ctlcmd.Run(context, strings.Fields(test.args), 0)
+		if test.exitCode > 0 {
+			unsuccessfulErr, ok := err.(*ctlcmd.UnsuccessfulError)
+			c.Assert(ok, Equals, true)
+			c.Check(unsuccessfulErr.ExitCode, Equals, test.exitCode)
+		} else {
+			if test.error == "" {
+				c.Check(err, IsNil)
+			} else {
+				c.Check(err, ErrorMatches, test.error)
+			}
+		}
+
+		c.Check(string(stdout), Equals, test.stdout, Commentf("%s\n", test.args))
+		c.Check(string(stderr), Equals, "")
+	}
+}
+
+func (s *isConnectedSuite) TestIsConnectedFromHook(c *C) {
+	s.st.Lock()
+	defer s.st.Unlock()
+
+	task := s.st.NewTask("test-task", "my test task")
+	setup := &hookstate.HookSetup{Snap: "snap1", Revision: snap.R(1), Hook: "test-hook"}
+
+	mockContext, err := hookstate.NewContext(task, s.st, setup, s.mockHandler, "")
+	c.Check(err, IsNil)
+
+	s.testIsConnected(c, mockContext)
+}
+
+func (s *isConnectedSuite) TestIsConnectedFromApp(c *C) {
+	s.st.Lock()
+	defer s.st.Unlock()
+
+	// ephemeral context
+	setup := &hookstate.HookSetup{Snap: "snap1", Revision: snap.R(1)}
+	mockContext, err := hookstate.NewContext(nil, s.st, setup, s.mockHandler, "")
+	c.Check(err, IsNil)
+
+	// sanity
+	c.Assert(mockContext.IsEphemeral(), Equals, true)
+
+	s.testIsConnected(c, mockContext)
+}
+
+func (s *isConnectedSuite) TestNoContextError(c *C) {
+	stdout, stderr, err := ctlcmd.Run(nil, []string{"is-connected", "foo"}, 0)
+	c.Check(err, ErrorMatches, `cannot check connection status without a context`)
+	c.Check(string(stdout), Equals, "")
+	c.Check(string(stderr), Equals, "")
+}
+
+func (s *isConnectedSuite) TestGetRegularUser(c *C) {
```

Comment:
so the only difference I can notice between this test and the `TestIsConnectedFromApp` is that we don't unlock state there, but we do here? Is that condition sufficient to emulate testing as a regular user?

Reply:
The key difference is in passing 0 vs 1000 (user uid) to ctlcmd.Run. Locking is present in both tests, just slightly different. I could get rid of TestGetRegularUser and make testIsConnected loop over the two uids if you think it's better? The only motivation to do this separately was to follow some other existing tests for this package and highlight the regular user aspect with a separate test (since we had issues in the past forgetting to update Run method to allow certain commands with non-root user).

Reply:
Ah yes, I didn't even notice that argument difference. I think doing this as a separate test like you have is fine, I just didn't quite follow the difference before. It would be a nice thing to do the locking the same I think, but it's a minor thing I think.